omnigibson/systems/macro_particle_system.py

```diff
@@ -812,92 +539,177 @@ def generate_group_particles_on_object(cls, group, max_samples, min_samples_for_
-        link_tfs = dict()
-        link_tfs_batch = np.zeros((cls.n_particles, 4, 4))
-        particle_local_poses_batch = np.zeros_like(link_tfs_batch)
-        particles = cls.particles if group is None else cls._group_particles[group]
-        for i, name in enumerate(particles):
-            link = cls._particles_info[name]["link"]
-            if link in link_tfs:
-                link_tf = link_tfs[link]
-            else:
-                link_tf = T.pose2mat(link.get_position_orientation())
-                link_tfs[link] = link_tf
-            link_tfs_batch[i] = link_tf
-            particle_local_poses_batch[i] = cls._particles_local_mat[name]
-
-        # Compute once
-        global_poses = np.matmul(link_tfs_batch, particle_local_poses_batch)
+                - (n, 3)-array: per-particle (x,y,z) position
+                - (n, 4)-array: per-particle (x,y,z,w) quaternion orientation
+        """
+        n_particles = len(particles)
+        if local:
+            global_poses = np.zeros((n_particles, 4, 4))
+            for i, name in enumerate(particles):
+                global_poses[i] = T.pose2mat(cls.particles[name].get_local_pose())
+        else:
+            # Iterate over all particles and compute link tfs programmatically, then batch the matrix transform
+            link_tfs = dict()
+            link_tfs_batch = np.zeros((n_particles, 4, 4))
+            particle_local_poses_batch = np.zeros_like(link_tfs_batch)
+            for i, name in enumerate(particles):
+                link = cls._particles_info[name]["link"]
+                if link in link_tfs:
+                    link_tf = link_tfs[link]
+                else:
+                    link_tf = T.pose2mat(link.get_position_orientation())
+                    link_tfs[link] = link_tf
+                link_tfs_batch[i] = link_tf
+                particle_local_poses_batch[i] = cls._particles_local_mat[name]
+
+            # Compute once
+            global_poses = np.matmul(link_tfs_batch, particle_local_poses_batch)
 
         # Decompose back into positions and orientations
         return global_poses[:, :3, 3], T.mat2quat(global_poses[:, :3, :3])
 
     @classmethod
-    def get_particle_position_orientation(cls, name):
+    def get_particles_position_orientation(cls, local=False):
+        return cls._compute_batch_particles_position_orientation(particles=cls.particles, local=local)
+
+    @classmethod
+    def get_group_particles_position_orientation(cls, group, local=False):
+        return cls._compute_batch_particles_position_orientation(particles=cls._group_particles[group], local=local)
+
+    @classmethod
+    def get_particle_position_orientation(cls, idx, local=False):
+        name = list(cls.particles.keys())[idx]
+        if local:
+            val = cls.particles[name].get_local_pose()
+        else:
+            # First, get local pose, scale it by the parent link's scale, and then convert into a matrix
```

Comment:
is this comment still accurate? how to handle scale here?

Reply:
Mention that `_particles_local_mat` already takes scale into account.

Reply:
done